pkg/backend/display/tree.go

```diff
@@ -168,22 +172,26 @@ func (r *treeRenderer) render(display *ProgressDisplay) {
 
 	// Convert system events into lines.
 	r.systemMessages = r.systemMessages[:0]
-	for _, payload := range display.systemEventPayloads {
+	for _, payload := range r.display.systemEventPayloads {
 		msg := payload.Color.Colorize(payload.Message)
 		r.systemMessages = append(r.systemMessages, splitIntoDisplayableLines(msg)...)
 	}
-
-	r.dirty = true
-	if r.opts.deterministicOutput {
-		r.frame(true, false)
-	}
 }
 
 func (r *treeRenderer) markDirty() {
 	r.m.Lock()
 	defer r.m.Unlock()
 
+	if r.display == nil || r.display.headerRow == nil {
+		// Don't mark dirty if there is no display, or
+		// if the display has never been initialized
+		return
+	}
+
```

Comment:
I have validated locally that this block will fix the whitespace issue  #15668 (and that its absence will cause it)